discover/service.go

```diff
@@ -0,0 +1,655 @@
+					nodePub, err)
+			}
+
+		// The discovery has been signalled to exit, to we exit our main
+		// loop so the wait group can be decremented.
+		case <-d.quit:
+			return
+		}
+	}
+}
+
+// processNetworkAnnouncement processes a new network relate authenticated
+// channel or node announcement. If the updates didn't affect the internal state
+// of the draft due to either being out of date, invalid, or redundant, then
+// false is returned. Otherwise, true is returned indicating that the caller
+// may want to batch this request to be broadcast to immediate peers during the
+// next announcement epoch.
+func (d *Discovery) processNetworkAnnouncement(aMsg *networkMsg) bool {
+	isPremature := func(chanID *lnwire.ChannelID) bool {
+		return chanID.BlockHeight > d.bestHeight
+	}
+
+	switch msg := aMsg.msg.(type) {
+
+	// A new node announcement has arrived which either presents a new
+	// node, or a node updating previously advertised information.
+	case *lnwire.NodeAnnouncement:
+		if aMsg.isRemote {
+			// TODO(andrew.shvv) Add node validation
+		}
+
+		node := &channeldb.LightningNode{
+			LastUpdate: time.Unix(int64(msg.Timestamp), 0),
+			Address:    msg.Address,
+			PubKey:     msg.NodeID,
+			Alias:      msg.Alias.String(),
+			AuthSig:    msg.Signature,
+		}
+
+		if err := d.cfg.Router.AddNode(node); err != nil {
+			log.Errorf("unable to add node: %v", err)
+			return false
+		}
+
+	// A new channel announcement has arrived, this indicates the
+	// *creation* of a new channel within the network. This only advertises
+	// the existence of a channel and not yet the routing policies in
+	// either direction of the channel.
+	case *lnwire.ChannelAnnouncement:
+		// If the advertised inclusionary block is beyond our knowledge
+		// of the chain tip, then we'll put the announcement in limbo
+		// to be fully verified once we advance forward in the chain.
+		if isPremature(&msg.ChannelID) {
+			blockHeight := msg.ChannelID.BlockHeight
+			log.Infof("Announcement for chan_id=(%v), is "+
+				"premature: advertises height %v, only height "+
+				"%v is known", msg.ChannelID, msg.ChannelID.BlockHeight,
+				d.bestHeight)
+
+			d.prematureAnnouncements[blockHeight] = append(
+				d.prematureAnnouncements[blockHeight],
+				aMsg,
+			)
+			return false
+		}
+
+		var proof *channeldb.ChannelAuthProof
+		if aMsg.isRemote {
+			// TODO(andrew.shvv) Add channel validation
+		}
+
+		proof = &channeldb.ChannelAuthProof{
+			NodeSig1:    msg.FirstNodeSig,
+			NodeSig2:    msg.SecondNodeSig,
+			BitcoinSig1: msg.FirstBitcoinSig,
+			BitcoinSig2: msg.SecondBitcoinSig,
+		}
+
+		edge := &channeldb.ChannelEdgeInfo{
+			ChannelID:   msg.ChannelID.ToUint64(),
+			NodeKey1:    msg.FirstNodeID,
+			NodeKey2:    msg.SecondNodeID,
+			BitcoinKey1: msg.FirstBitcoinKey,
+			BitcoinKey2: msg.SecondBitcoinKey,
+			AuthProof:   proof,
+		}
+
+		if err := d.cfg.Router.AddEdge(edge); err != nil {
+			if !routing.IsError(err, routing.ErrOutdatedEdge) {
```

Comment:
In previous version we logged the `outdated` announcement error with `info` level.